Use ProbeTable to generate NP probes information

chrishalcrow opened the pull request:
PR to use ProbeTable (https://github.com/billkarsh/ProbeTable/) to generate NeuroPixels probe information, instead of hardcoding this ourselves.

Import differences in `src/probeinterface/neuropixels_tools.py`

Does a few things:
- Replaces the `npx_descriptions` hardcoded dict with a `make_npx_descriptions` function which generates this data from the `probe_features.json` file from ProbeTable.
- We now extract a `mux_table_array` from the `probe_features.json`, file, which tells us the adc groupings. This can then be used in spikeinterface for the `inter_sample_shift`. PR incoming. This is saved as a probe annotation.
- _Almost_ deprecates the `probe_type` argument. Sticky point is the read/write imro function, where `probe_type` is stored in the format. E.g. `tests/data/imro/test_single_shak_2.0.imro` starts “(21,384)(…” and the “21” is the `probe_type` of the probe.
- REMOVES `ptype` as a probe annotation 
- Adds a `get_probe_contour_vertices` function which computes the probe contour vertices from metadata.


Implementation questions and annoying edge cases:
- ProbeTable doesn’t contain what we called the `Phase3a` probe. This is _almost_ a NP1010 but doesn’t contain the `ap_hipas_flt` field in the imro table. So we have to deal with it separately. It also has “probe_part_number = None" which is… not so nice. It’s hard to tell if a read in probe is unknown or is this one.
- We had a `model_name` for each probe, which ProbeTable doesn’t contain. @alejoe91 suggested using the `description`, which changes all our probe names. Ok??
- I’ve saved the `json` file at `src/probeinterface/resources/probe_features.json`. I think it has to be in source to work for a normal pip installation? Any better places to store?
- We want to check backwards compatibility, but this PR most changes the `npx_description` stuff so it's too radical. So we only really need check that these match. To do this, I’ve moved the old hardcoded dict to a test file. Any other ideas for how to check this? There are some mismatches between the ProbeTable and old ProbeInterface implementation. These are:

|Probe type|key|ProbeTable|ProbeInterface|
|---|---|---|---|
|1020|x_pitch|87|91|
||stagger|16|12|
|1021|x_pitch|87|91|
||stagger|16|12|
|1030|x_pitch|87|56|
||stagger|16|12|
|1031|x_pitch|87|91|
||stagger|16|12|
|1121|x_pitch|0|6|
||contact_width|2.5|2|
|1110|fields_in_imro_table|(“group”, “bankA”, “bankB”)|(“channel_ids”, “banks”, “references”, “ap_gains”, “lf_gains”, “ap_hp_filters”|

We checked these, and trust ProbeTable.

To do:
- Write a GitHub action which copies `probe_features.json` from ProbeTable when it's updated.
- Add more details in docstrings.

Just wanted to get some feedback/ideas before going any further!

h-mayorquin commented:
Thanks for taking on this, @chrishalcrow 

> I’ve saved the json file at src/probeinterface/resources/probe_features.json. I think it has to be in source to work for a normal pip installation? Any better places to store?

Yes, you will need to modify the pyproject.toml to ensure that this is packaged on the wheel/stds. An alternative is to store it as a python dict which will make package simpler (that way you don't need to modify the gitignore).

All that said, json should be fine as well, any reason you prefer it? Whatever we choose, we should add a cron test that runs weekly or monthy to ensure that what we have matches:

https://github.com/billkarsh/ProbeTable/blob/main/Tables/probe_features.json

I can help with that if you want.

h-mayorquin commented on src/probeinterface/neuropixels_tools.py:
```diff
@@ -515,10 +57,208 @@
+def get_probe_contour_vertices(shank_width, tip_length, probe_length) -> list:
+    """
+    Function to get the vertices of the probe contour from probe properties. The probe contour can be constructed
+    from five points. These are the vertices shown in the following figure:
```
Thanks for the diagram.

h-mayorquin commented on src/probeinterface/neuropixels_tools.py:
```diff
@@ -515,10 +57,208 @@
+    # Construct probe contour, for styling the probe
+    shank_width = float(pt_metadata["shank_width_um"])
+    tip_length = float(pt_metadata["tip_length_um"])
+    probe_length = 10_000
```
? any reason this is 10_000

h-mayorquin commented on src/probeinterface/neuropixels_tools.py:
```diff
@@ -515,10 +57,208 @@
+
+    is_phase3a = False
+    # These are all prototype NP1.0 probes, not contained in ProbeTable
+    if probe_part_number in ["PRB_1_4_0480_1", "PRB_1_4_0480_1_C", "PRB_1_2_0480_2", None]:
```
`PRB_1_4_0480_1` and `PRB_1_4_0480_1_C` are on the table, why are we fetching `NP1010` instead for those?

h-mayorquin commented:
> We had a model_name for each probe, which ProbeTable doesn’t contain. @alejoe91 suggested using the description, which changes all our probe names. Ok??

I think this is a good idea.I am +1 on it.

h-mayorquin commented on src/probeinterface/neuropixels_tools.py:
```diff
@@ -515,10 +57,208 @@
+    probe_length = 10_000
+    pi_metadata["contour_description"] = get_probe_contour_vertices(shank_width, tip_length, probe_length)
+
+    # Get the mux table
```
Ah, this is the multiplexing units so they are used for phase shift. I still think that we should add acomment.

chrishalcrow commented on src/probeinterface/neuropixels_tools.py:
```diff
@@ -633,22 +372,21 @@ def _read_imro_string(imro_str: str, imDatPrb_pn: Optional[str] = None) -> Probe
     imro_table_header_str, *imro_table_values_list, _ = imro_str.strip().split(")")
     imro_table_header = tuple(map(int, imro_table_header_str[1:].split(",")))
 
+    imDatPrb_type = None
```
Hello, I think we *do* need the type for when we read `.imro` files directly. This is how `test_single_shake_2.0.imro` starts:

```(21,384)(0 1 0 0)(1 1 0 1)(2 1 0 2)(3 1 0 ....```

(the rest of the file is just 4 digit numbers in brackets)

and `21` is the probe_type. So we need to be able to figure out which `probe_part_number` corresponds to this type. So we either keep a dict, or we write a function that tries to figure out the `probe_part_number` from the ProbeTables file. I can't see an easy way to make this function. The types are in e.g. `probe_features['z_imro_formats']['imro_np1000_val_def']` and this one equals `type:{0,1020,1030,1100,1120,1121,1122,1123,1200,1300}`. If it's any of the 4-digit numbers, the probe part number is "NP"+number. If it's "0", it's either "NP1010", "PRB_1_4_0480_1", "PRB_1_4_0480_1_C" or "PRB_1_2_0480_2". The others are easier.

chrishalcrow commented on src/probeinterface/neuropixels_tools.py:
```diff
@@ -515,10 +57,208 @@
+
+    probe_features_filepath = Path(__file__).absolute().parent / Path("resources/probe_features.json")
+    probe_features = json.load(open(probe_features_filepath, "r"))
+    pt_metadata = probe_features["neuropixels_probes"].get(probe_part_number)
```
Yes. Should I document it the first time I write it? Or do you have a better notation idea?

h-mayorquin commented:
The logic seems fine and the tests are passing. We can improve later if we need to so this is fine to me. To points left are that:
* The packaging works. We should see if the json artifact is packed on the sdist and wheels for distribution.
* See that if things differ with the current https://github.com/SpikeInterface/probeinterface_library/tree/main/neuropixels they should differ for a good reason.

chrishalcrow commented:
Great! I've...
- added more backwards compatibility information in the PR description.
- added a github action (99% written by Claude) which checks the `probe_features.json` every week. It works fine on `chrishalcrow/probeinterface` but I'm unsure how to test it here.

Not sure how to check the packaging stuff - we can discuss offline.

chrishalcrow commented on src/probeinterface/neuropixels_tools.py:
```diff
@@ -630,65 +340,127 @@ def _read_imro_string(imro_str: str, imDatPrb_pn: Optional[str] = None) -> Probe
+    probe = _make_npx_probe_from_description(pt_metadata, elec_ids, shank_ids, mux_table)
 
-    # this is scalar annotations
+    mux_table_format_type = pt_metadata["mux_table_format_type"]
```
Should delete these three lines, and the line `mux_table_array=mux_table_array,` below

samuelgarcia commented on tests/test_neuropixels_backwards_compat.py:
```diff
@@ -0,0 +1,495 @@
+from probeinterface.neuropixels_tools import make_npx_description, probe_part_number_to_probe_type
+
```
```suggestion
# this code is here only for testing the refactoring of neuropixel probe handling done by Chris in 2025 and based on .....
# this test could be removed in 202X ....

```

samuelgarcia replied:
filenally remove this file

samuelgarcia commented on tests/test_neuropixels_backwards_compat.py:
```diff
@@ -0,0 +1,495 @@
+
+        if probe_part_number not in ["3000", "1200"]:
+
+            probe_info = make_npx_description(probe_part_number)
```
finally no too complex

samuelgarcia commented:
This is OK for me for merging.

@alejoe91 @chrishalcrow I guess that the mux array annotations will be used in get_neuropixels_sample_shifts on spikeinterface side.
Should we make the PR before merging it ?

samuelgarcia commented:
ET voilà!
